README.md
# swagger-stub
Generate a stub from a swagger file


HISTORY.rst
=======
History
=======

0.2.1 (2016-6-6)
------------------

* Fix repeated base path.

0.2.0 (2016-5-2)
------------------

* Add access to definition example.

0.1.1 (2016-1-31)
------------------

* Change license to MIT.

0.1 (2016-1-29)
------------------

* First release on PyPI.


tests/test_swagger_stub.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import json
import os
import pytest
import requests

from swagger_stub import swagger_stub


@pytest.fixture
def swagger_stub_test():
    try:
        return swagger_stub([(os.path.join(os.path.dirname(__file__), 'swagger.yaml'), 'http://localhost:8000')]).next()
    except AttributeError:
        return swagger_stub([(os.path.join(os.path.dirname(__file__), 'swagger.yaml'), 'http://localhost:8000')]).__next__()


@pytest.fixture
def pet_definition_example():
    return {
        'category': {
            'id': 42,
            'name': 'string'
        },
        'status': 'string',
        'name': 'doggie',
        'tags': [
            {
                'id': 42,
                'name': 'string'
            }
        ],
        'photoUrls': [
            'string',
            'string2'
        ],
        'id': 42
    }


def test_swagger_stub(swagger_stub_test, pet_definition_example):
    # Test data example
    swagger_stub_test.definitions['Pet'] == pet_definition_example

    # Wrong url
    response = requests.get('http://localhost:8000/v2/error')
    assert response.status_code == 404
    response = requests.get('http://localhost:8000/pets/5121')
    assert response.status_code == 404
    response = requests.post('http://localhost:8000/v2/error')
    assert response.status_code == 400

    # Wrong body type
    response = requests.post('http://localhost:8000/v2/pets', data='error')
    assert response.status_code == 400
    assert response.json() == {'body': ['Not valid json.']}

    # Post
    response = requests.post('http://localhost:8000/v2/pets', data=json.dumps(pet_definition_example))
    assert response.status_code == 201

    # Wrong body
    response = requests.post('http://localhost:8000/v2/pets', data=json.dumps({}))
    assert response.status_code == 400

    # Request with path parameter
    response = requests.get('http://localhost:8000/v2/pets/5121')
    assert response.status_code == 200
    assert response.json() == pet_definition_example

    # Mock
    swagger_stub_test.add_mock_call('get', '/v2/test', {'mock': 'call'})
    response = requests.get('http://localhost:8000/v2/test')
    assert response.json() == {'mock': 'call'}

    # Side effect
    swagger_stub_test.add_mock_side_effect('get', '/v2/side', [{'test': '1'}, {'test': '2'}, {'test': '3'}])
    response = requests.get('http://localhost:8000/v2/side')
    assert response.json() == {'test': '1'}
    response = requests.get('http://localhost:8000/v2/side')
    assert response.json() == {'test': '2'}
    response = requests.get('http://localhost:8000/v2/side')
    assert response.json() == {'test': '3'}

    swagger_stub_test.add_mock_side_effect('get', '/v2/error', Exception)

    with pytest.raises(Exception):
        response = requests.get('http://localhost:8000/v2/error')


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'Trax-Air/swagger_stub'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: {0!s})'.format(GITHUB_REPO))
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


swagger_stub/swagger_stub.py
# -*- coding: utf-8 -*-

import httpretty
import inspect
import json
import pytest
import re

try:
    from urlparse import parse_qsl
    from urlparse import urlsplit
except ImportError:  # Python3
    from urllib.parse import parse_qsl
    from urllib.parse import urlsplit

from swagger_parser import SwaggerParser


class StubMemory(object):
    """Store data about calls made to the stub."""

    def __init__(self, swagger_parser):
        self.memory = []
        self.mock_call = {}
        self.side_effect = {}

        self.definitions = swagger_parser.definitions_example

    def add_call(self, action, path, body, query, status_code):
        """Add a call to the memory

        Args:
            action: action of the call.
            path: path of the call.
            body: body of the call.
            query: query of the call.
            status_code: status_code of the call.
        """
        self.memory.append({'path': path,
                            'action': action,
                            'body': body,
                            'query': query,
                            'status_code': status_code})

    def get_call(self, action=None, path=None):
        """Get a call with the given action and/or path.

        Args:
            action: action of the call.
            path: path of the call.
        """
        return_call = []
        for data in self.memory:
            if action is None or data['action'] == action and\
               path is None or data['path'] == path:
                return_call.append(data)

        return return_call

    def add_mock_call(self, action, path, body, status_code=200):
        """Add a mock call to the list.

        Args:
            action: action of the call to mock.
            path: path of the call to mock.
            body: body to return during the call.
        """
        if action not in self.mock_call:
            self.mock_call[action] = {}
        self.mock_call[action][path] = (body, status_code)

    def add_mock_side_effect(self, action, path, side_effect):
        """Add a side effect to a call.

        Args:
            action: action of the call to mock.
            path: path of the call to mock.
            side_effect: either an Exception, a function or a [(body, status_code)] or [body]
        """
        if action not in self.side_effect:
            self.side_effect[action] = {}
        self.side_effect[action][path] = {}
        self.side_effect[action][path]['counter'] = -1
        self.side_effect[action][path]['effect'] = side_effect

    def process_side_effect(self, action, path):
        """Get the side_effect for the given action and path.

        Args:
            action: http action.
            path: path of the request to mock.

        Returns:
            (body, status_code) or raise the side_effect error

        Raises:
            IndexError: We have gone through the side_effect list (if it's a list)
        """
        # TODO(Handle functions)
        if action in self.side_effect and path in self.side_effect[action]:
            if (inspect.isclass(self.side_effect[action][path]['effect']) and
                issubclass(self.side_effect[action][path]['effect'], Exception)) or \
                    (inspect.isclass(type(self.side_effect[action][path]['effect'])) and
                        issubclass(type(self.side_effect[action][path]['effect']), Exception)):
                raise self.side_effect[action][path]['effect']
            elif isinstance(self.side_effect[action][path]['effect'], list):
                # Update the counter
                self.side_effect[action][path]['counter'] += 1
                index = self.side_effect[action][path]['counter']
                if isinstance(self.side_effect[action][path]['effect'][index],
                              tuple):
                    return (json.dumps(self.side_effect[action][path]['effect'][index][0]),
                            self.side_effect[action][path]['effect'][index][1])
                else:
                    return json.dumps(self.side_effect[action][path]['effect'][index]), 200


swagger_url = {}  # Store SwaggerParser with the associate url
memory = {}  # Stub memory


def get_parsed_body(request):
    """Get the body (json or formData) of the request parsed.

    Args:
        request: request object to get the body from.

    Returns:
        A dictionary of the body.

    Raises:
        ValueError: if it is neither a json or a formData
    """
    data = None
    if not request.parsed_body == '':
        try:
            data = json.loads(request.body.decode('utf-8'))
        except Exception:  # Decode formData
            data = dict(parse_qsl(request.body))

            if not data:  # Not a form data
                raise
    return data


def get_data_from_request(request, uri, headers):
    """Get the data for the given request.

    Args:
        request: the request to process.
        uri: uri of the request.
        headers: headers of the request.

    Returns:
        (status code, headers, reponse message)
    """
    # Get body data
    data = None
    try:
        data = get_parsed_body(request)
    except ValueError:  # Decode formData
        return 400, headers, '{"body": ["Not valid json."]}'

    # Get uri data
    data_url = urlsplit(uri)
    mem = memory['{uri.scheme}://{uri.netloc}'.format(uri=data_url)]
    swagger_parser = swagger_url[
        '{uri.scheme}://{uri.netloc}'.format(uri=data_url)]

    path = data_url[2]
    query = dict(parse_qsl(data_url[3]))
    action = request.method.lower()

    # Get swagger spec
    if path == '/swagger.json':
        return 200, headers, swagger_parser.json_specification

    # Check mock call
    if action in mem.mock_call and path in mem.mock_call[action]:
        mem.add_call(action, path, data, query, mem.mock_call[action][path][1])
        return mem.mock_call[action][path][1], headers, json.dumps(mem.mock_call[action][path][0])

    # Check side_effect
    side = mem.process_side_effect(action, path)
    if side is not None:
        return (side[1], headers, side[0])

    # Test path + action exists
    if swagger_parser.get_path_spec(path, action) == (None, None):
        if action == 'post':
            mem.add_call(action, path, data, query, 400)
            return 400, headers, ''  # Invalid operation
        else:
            mem.add_call(action, path, data, query, 404)
            return 404, headers, ''  # Not found

    # Test request
    if swagger_parser.validate_request(path, action, data, query):
        # Return response from SwaggerParser
        response = swagger_parser.get_request_data(path, action)
        mem.add_call(action, path, data, query, min(response.keys()))
        return (min(response.keys()), headers, json.dumps(response[min(response.keys())]))
    else:
        # Invalid request
        mem.add_call(action, path, data, query, 400)
        return (400, headers, '')


@pytest.yield_fixture
def swagger_stub(swagger_files_url):
    """Fixture to stub a microservice from swagger files.

    To use this fixture you need to define a swagger fixture named
    swagger_files_url with the path to your swagger files, and the url to stub.
    Then just add this fixture to your tests and your request pointing to the
    urls in swagger_files_url will be managed by the stub.

    Example:
        @pytest.fixture
        def swagger_files_url():
            return [('tests/swagger.yaml', 'http://localhost:8000')]
    """
    httpretty.enable()

    for i in swagger_files_url:  # Get all given swagger files and url
        base_url = i[1]
        s = SwaggerParser(i[0])
        swagger_url[base_url] = s

        # Register all urls
        httpretty.register_uri(
            httpretty.GET, re.compile(base_url + r'/.*'),
            body=get_data_from_request)

        httpretty.register_uri(
            httpretty.POST, re.compile(base_url + r'/.*'),
            body=get_data_from_request)

        httpretty.register_uri(
            httpretty.PUT, re.compile(base_url + r'/.*'),
            body=get_data_from_request)

        httpretty.register_uri(
            httpretty.PATCH, re.compile(base_url + r'/.*'),
            body=get_data_from_request)

        httpretty.register_uri(
            httpretty.DELETE, re.compile(base_url + r'/.*'),
            body=get_data_from_request)

        memory[base_url] = StubMemory(s)
        yield memory[base_url]

    # Close httpretty
    httpretty.disable()
    httpretty.reset()
